2016/proposals/result-aggregation-server-garg-prerit.md

```diff
@@ -0,0 +1,404 @@
+from flask.ext.script import Manager
+from flask.ext.migrate import Migrate, MigrateCommand
+
+app = Flask(__name__)
+app.config['SQLALCHEMY_DATABASE_URI'] = 'sqlite:///app.db'
+
+db = SQLAlchemy(app)
+migrate = Migrate(app, db)
+
+manager = Manager(app)
+manager.add_command('db', MigrateCommand)
+
+if __name__ == '__main__':
+    manager.run()
+```
+
+### Frontend
+* For the Server Administrator, a basic **frontend** can be developed for them to manage and analyse the results, e.g. 
+ * Common install problems faced in a particular workshop 
+ * Common install problems faced by students overall. 
+ * Which version of package is most prone to errors, etc. 
+  
+* The front end can be developed using simple HTML, CSS, [Bootstrap](http://getbootstrap.com/) , Javascript.  
+* For the greater good, some plots can be generated by extracting useful information about the failures. This can be done using [Plotly](https://plot.ly/). This will provide better understanding by visualization. Code Snippet :
+```python
+import plotly.plotly as py
+import plotly.graph_objs as go
+
+data = [
+            go.Bar(
+                x=['Ubuntu' , 'Redhat' , 'Fedora'],
+                y=[30 , 20 , 40] ,
+            )
+        ]
+        file_name = "filename"
+        plot_url = py.plot(data, filename=file_name , auto_open=False)
+```
+
+### Examples of results and Plots :
+* Figuring out the most error prone package during installation
+![install_plot](http://i.imgur.com/z85Nx40.png)
+* Figuring out the most error prone Operating system and environment in installation of a particular package.
+![os_plot](http://i.imgur.com/bajsojc.png)
+
+### API Structure
+* `run.py` : Server shall be run from this file.
+* `manage.py`: Migration script to perform database migration using Flask-migrate.
+* `config.py` : Application directory, SQLALCHEMY_DATABASE_URI , Secret keys etc shall be put in this file.
+* `models.py` : All the models shall be defined using SQLAlchemy in this file.
+* `views.py` : All the controllers, API endpoints shall be defined here.
+* `__init__.py` : Defining the database object, database creation shall be done here.
+* `templates` : The front end HTML files shall be kept here.
+* `static` : The Bootstrap, js, CSS files, images (if any) etc shall be kept here.
+
+**Rough Directory Structure** :
+```
+~/Installation_server
+    |-- run.py
+    |-- manage.py
+    |-- config.py
+    |__ /env             # Virtual Environment
+    |__ /app             # Our Application        
+         |-- __init__.py
+         |-- views.py
+         |-- models.py
+         |__ /templates
+             |-- 404.html
+             |__ /auth
+                 |-- signin.html
+         |__ /static
+```
+**Tentative python modules** to be used :
+`flask`, `flask-SQLAlchemy`, `Jinja2`, `SQLAlchemy`, `flask-migrate`, `flask-login` , 'platform', `plotly.plotly`, `plotly.graph_objs`, `requests`, `datetime`, `os`, `unittest` , `logging` etc..
```

Comment:
I just realized that you cannot rely on `requests` since it would have to be installed on the user computer (it may not be).